### PR TITLE
add support for managed thread pool

### PR DESCRIPTION
**Pull request by ajwootto**

### CODE REVIEW

**Comment by ajwootto** [comment]
Current dependencies on/for this PR:
* main
  * **PR #101** <a href="https://app.graphite.dev/github/pr/DevCycleHQ/go-server-sdk/101" target="_blank">Graphite</a> 
    * **PR #102** <a href="https://app.graphite.dev/github/pr/DevCycleHQ/go-server-sdk/102" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/DevCycleHQ/go-server-sdk/102?utm_source=stack-comment).

File: client_test.go [hunk 1]
```diff
@@ -165,3 +166,100 @@ func BenchmarkDVCClient_Variable(b *testing.B) {
...
+	b.ResetTimer()
+	b.ReportAllocs()
+	var i int
+	for i = 0; i < b.N; i++ {
+		wg.Add(1)
+
+		go func() {
+			defer wg.Done()
+			variable, err := client.Variable(user, test_large_config_variable, false)
+			//if i%200 == 1 {
```
Comment (JamieSinn) [inline client_test.go:202]: why?

File: client_test.go [hunk 2]
```diff
@@ -165,3 +166,100 @@ func BenchmarkDVCClient_Variable(b *testing.B) {
...
```
Comment (JamieSinn) [inline client_test.go:252]: por que?

File: client_test.go (hunk 1 above)
Comment (ajwootto) [reply client_test.go:202]: was testing what happened when setting config while threads were busy executing variable calls

File: configuration.go [hunk 3]
```diff
@@ -98,6 +115,10 @@ func (o *DVCOptions) CheckDefaults() {
 	} else if o.MaxMemoryAllocationBuckets <= -1 {
 		o.MaxMemoryAllocationBuckets = 0
 	}
+
+	if o.MaxWasmWorkers <= 0 {
+		o.MaxWasmWorkers = runtime.NumCPU()
```
Comment (JamieSinn) [inline configuration.go:120]: This should be GOMAXPROCS

File: client.go [hunk 4]
```diff
@@ -54,7 +57,6 @@ type service struct {
 }
 
 func initializeWasmMain(options *DVCOptions) (ret *WASMMain, err error) {
-	options.CheckDefaults()
```
Comment (JamieSinn) [inline client.go:57]: is this being removed or moved somewhere else?

**Review by JamieSinn — CHANGES_REQUESTED** [review]

**Review by JamieSinn — APPROVED** [review]

**Review by chris-hoefgen — APPROVED** [review]
These look good to me and combined with the load test i'm feeling good about the changes. I would give Rob the final approval as his Go skillz are stronger than me.

File: client.go [hunk 5]
```diff
@@ -609,6 +646,10 @@ func (c *DVCClient) Close() (err error) {
 		<-c.internalOnInitializedChannel
 	}
 
+	if c.bucketingWorkerPool != nil {
+		c.bucketingWorkerPool.Close()
+	}
```
Comment (rob-odwyer) [inline client.go:651]: Have we tested this Close flow anywhere? I don't think we used it in the benchmark(s) at all.

**Review by rob-odwyer — COMMENTED** [review]
Couple of questions, but looks pretty close to ready!

File: client.go [hunk 6]
```diff
@@ -585,8 +603,27 @@ func (c *DVCClient) SetClientCustomData(customData map[string]interface{}) error
...
 			}
-			err = c.localBucketing.SetClientCustomData(string(data))
-			return err
+			err = c.localBucketing.SetClientCustomData(data)
+
+			if err != nil {
+				return err
+			}
+
+			if c.bucketingWorkerPool != nil {
```
[the remark this answers is not in the document]
Comment (ajwootto) [reply client.go:612]: we could consider just always using the pool even when the CPU count is 1? that would eliminate some redundant code paths. I originally did it this way to de-risk adding the thread pooling

File: client.go (hunk 5 above)
Comment (ajwootto) [reply client.go:651]: it's tested by the test harness

File: thread_pool.go [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+	VariableEvalParams *[]byte
+	ConfigData         *[]byte
+	ClientCustomData   *[]byte
+	VariableType       VariableTypeCode
+}
+
+type WorkerPoolResponse struct {
+	Variable *proto.SDKVariable_PB
+	Events   *PayloadsAndChannel
+	ImHere   bool
```
[the remark this answers is not in the document]
Comment (ajwootto) [reply thread_pool.go:51]: forgot to remove 👍

File: client.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (rob-odwyer) [reply client.go:612]: That's probably the simplest approach, but I think we can leave it for a later refactor.

**Pull request merged**